server/src/auth.middleware.ts

```diff
@@ -15,15 +20,52 @@ export class AuthMiddleware implements NestMiddleware {
 
     req.session = false;
 
-    const { authorization = '' } = req.headers;
+    const { authorization = '', referer } = req.headers;
     const token = authorization.split(' ')[1];
 
     try {
-      req.session = await this.authService.validateCurrentToken(token);
+      // this promise will throw if invalid
+      let validatedToken = await this.authService.validateCurrentToken(token);
+
+      // the referer is the domain issuing the request. in this case, it'll be the client domain,
+      if (referer) {
+        // which includes the query params
+        const refererParams = new URL(referer);
+        const email = refererParams.searchParams.get('email'); // the query param, email, sent from the client. this is who the "Creeper" wants to be.
+        const { mail } = validatedToken; // the "creepers" actual email, for verification.
+
+        // REDO: env variables.
+        // if an e-mail is provided, implicitly it means force creeper mode. then verify creeper mode
+        // with some criteria.
+        if (email && (mail === 'dcpcreeper@gmail.com' || mail.includes('@planning.nyc.gov'))) {
+          validatedToken = await this._spoofToken(validatedToken, email);
+        }
+      }
+
+      req.session = validatedToken;
 
       next();
     } catch (e) {
       next();
     }
   }
+
+  // spoof a token for the creeper user, provided their verified token and email param
+  // returns a new token with new credentials, allowing all routes access to the spoofed user's
+  // resources
+  async _spoofToken(validatedToken, creeperEmail) {
+    const NYCID_TOKEN_SECRET = this.config.get('NYCID_TOKEN_SECRET');
+
+    // these simulate the flow of authentication for the app
+    const spoofedNycIdToken = jwt.sign({
+      ...validatedToken,
+      mail: creeperEmail,
+    }, NYCID_TOKEN_SECRET);
+    const spoofedZapToken = await this.authService.generateNewToken(spoofedNycIdToken);
+    const validatedSpoofedToken = await this.authService.validateCurrentToken(spoofedZapToken);
```

Comment:
I always wondered -- why do we need to validate a token we just signed?

Reply:
@godfreyyeung yeah, good question. I think it's not quite the best name for the method. but the jwt method that's invoked there both validates and _parses/decrypts_ the token into an object.